federation/pkg/federation-controller/configmap/configmap_controller.go

```diff
@@ -0,0 +1,328 @@
+			return cache.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						versionedOptions := util.VersionizeV1ListOptions(options)
+						return targetClient.Core().ConfigMaps(api_v1.NamespaceAll).List(versionedOptions)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						versionedOptions := util.VersionizeV1ListOptions(options)
+						return targetClient.Core().ConfigMaps(api_v1.NamespaceAll).Watch(versionedOptions)
+					},
+				},
+				&api_v1.ConfigMap{},
+				controller.NoResyncPeriodFunc(),
+				// Trigger reconciliation whenever something in federated cluster is changed. In most cases it
+				// would be just confirmation that some configmap opration succeeded.
+				util.NewTriggerOnAllChanges(
+					func(obj pkg_runtime.Object) {
+						configmapcontroller.deliverConfigMapObj(obj, configmapcontroller.configmapReviewDelay, false)
+					},
+				))
+		},
+
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *federation_api.Cluster) {
+				// When new cluster becomes available process all the configmaps again.
+				configmapcontroller.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(configmapcontroller.clusterAvailableDelay))
+			},
+		},
+	)
+
+	// Federated updeater along with Create/Update/Delete operations.
+	configmapcontroller.federatedUpdater = util.NewFederatedUpdater(configmapcontroller.configmapFederatedInformer,
+		func(client kubeclientset.Interface, obj pkg_runtime.Object) error {
+			configmap := obj.(*api_v1.ConfigMap)
+			_, err := client.Core().ConfigMaps(configmap.Namespace).Create(configmap)
+			return err
+		},
+		func(client kubeclientset.Interface, obj pkg_runtime.Object) error {
+			configmap := obj.(*api_v1.ConfigMap)
+			_, err := client.Core().ConfigMaps(configmap.Namespace).Update(configmap)
+			return err
+		},
+		func(client kubeclientset.Interface, obj pkg_runtime.Object) error {
+			configmap := obj.(*api_v1.ConfigMap)
+			err := client.Core().ConfigMaps(configmap.Namespace).Delete(configmap.Name, &api_v1.DeleteOptions{})
+			return err
+		})
+	return configmapcontroller
+}
+
+func (configmapcontroller *ConfigMapController) Run(stopChan <-chan struct{}) {
+	go configmapcontroller.configmapInformerController.Run(stopChan)
+	configmapcontroller.configmapFederatedInformer.Start()
+	go func() {
+		<-stopChan
+		configmapcontroller.configmapFederatedInformer.Stop()
+	}()
+	configmapcontroller.configmapDeliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		configmap := item.Value.(*configmapItem)
+		configmapcontroller.reconcileConfigMap(configmap.namespace, configmap.name)
+	})
+	configmapcontroller.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
+		configmapcontroller.reconcileConfigMapsOnClusterChange()
+	})
+	util.StartBackoffGC(configmapcontroller.configmapBackoff, stopChan)
+}
+
+func getConfigMapKey(namespace, name string) string {
```

Comment:
As per previous reviews, perhaps just use NamespacedName.String() to avoid redundancy?